Add mixed affinity `UPDATE`, unify UUID behaviour

mkgrgis opened the pull request:
In this PR
- `UPDATE` works with `column_type` option for mixed affinity cases and writes different affinity values depends on value of the option. This fix https://github.com/pgspider/sqlite_fdw/issues/90. This issue affects `uuid` and `timestamp` types of data, implemented in different time. Mixed affinity `timestamp` was implemented before `uuid` support in https://github.com/pgspider/sqlite_fdw/pull/82,
- UUID behaviour and tests are unified between all supported PostgreSQL versions for full mixed affinity support with both of `SELECT`/`WHERE` usage, `INSERT` and `UPDATE`,
- regress tests reordered: `sqlite_fdw_post` firsts, than tests about data types in ABC order, other sequence with no changes,
- add a function which converted  `column_type` option value to SQLite affinity code for standard values with `int` as `integer` affinity synonym,
- `uuid` test renumbering and add `COSTS OFF` in some TCs,
- add to `uuid` test important TC about semantics conflict in SQLite: the same UUID values with different affinity, but no error during SQLite unique check,
- add current behaviour for `Infinity` values in `floatN` column into `type` tests with SQLite unique value check. No code changes. This behaviour should be tested _before_ special planned PR about special values in `floatN` columns,
- add translation rule for `IMPORT FOREIGN SCHEMA` if SQLite formal column  name is `uuid`.

mkgrgis commented:
@t-kataym, this PR was rebased to current `master` branch. Could you please appoint a review?

t-kataym commented:
@mkgrgis , Thank you for your contribution. We would like to confirm the PR and then merge it in this month if no problem.

son-phamngoc commented:
Hello @mkgrgis , thanks for your hard work.
I have 2 questions about this description. Could you help me to answer them?

> add translation rule for IMPORT FOREIGN SCHEMA if SQLite formal column name is uuid.

I see that you created a custom function to convert text/blob to uuid, and deparsed remote UPDATE query to call it. Therefore, I think this change should be for UPDATE, not IMPORT FOREIGN SCHEMA. Is that correct?

```
special_affinity = (pg_attyp == UUIDOID && preferred_affinity == SQLITE3_TEXT) ||
                   (pg_attyp == TIMESTAMPOID && preferred_affinity == SQLITE_INTEGER);
if (special_affinity)
{
    ....
```
I see that you handled 2 cases here: uuid and timestamp, but your description in this PR only mentioned about uuid case.  
Could you add more description for timestamp case?

mkgrgis commented:
Thanks, @son-phamngoc ! Added in description 

>  fix https://github.com/pgspider/sqlite_fdw/issues/90. This issue affects `uuid` and `timestamp` types of data, implemented in different time. Mixed affinity `timestamp` was implemented before `uuid` support in https://github.com/pgspider/sqlite_fdw/pull/82,

son-phamngoc commented:
> Yes. Rule for IMPORT FOREIGN SCHEMA is outside of deparsing. It's only about better table auto DDL for PostgreSQL. In SQLite was no strict column data types or affinity before special CREATE TABLE ...STRICT mode.

@mkgrgis So, I think the description should be "add translation rule for UPDATE if SQLite formal column name is uuid.". Could you update it?

mkgrgis commented:
> So, I think the description should be "add translation rule for UPDATE if SQLite formal column name is uuid

No, @son-phamngoc. In SQLite data processing code there is a common good style not to use formal column definitions at all, only _affinity de facto_ of a SQLite data value. `UPDATE`, `SELECT`, any `WHERE` etc. works only based on affinity.
In this FDW common code style is not applicable only to a function for `IMPORT FOREIGN SCHEMA`. In my PR there is no formal column definition usage outside of the function. Why? Because `IMPORT FOREIGN SCHEMA` gives recommended data types for PostgreSQL based on SQLite formal column definitions. FDW thinks the result will be normal, but is ready to any affinity according https://github.com/pgspider/sqlite_fdw/tree/master?tab=readme-ov-file#datatypes .

son-phamngoc commented:
@mkgrgis I thought that description related to your modification in `sqlite_deparse_direct_update_sql` function of deparse.c.  
But according to your explanation, it seems it is different.  
Could you point me out which parts of your modification related to that description? I do not see any code change or test case related to IMPORT FOREIGN SCHEMA in this PR.

mkgrgis commented:
> Could you point me out which parts of your modification related to that description?

I am sorry, @son-phamngoc. There was no that modification. I was sure this was committed by me, but before https://github.com/pgspider/sqlite_fdw/pull/95/commits/9c81b0fa941c1e56b4c6badcb5820c16c6e5581b this was only in my private draft. Now this is fixed.

son-phamngoc commented on sqlite_data_norm.c:
```diff
@@ -142,21 +143,68 @@ sqlite_fdw_uuid_blob (const unsigned char* s0, unsigned char* Blob)
+			return;
+		}
+	}
+	if (t != SQLITE_BLOB)
```
I think the logic is not correct for TEXT type.  
If `t == SQLITE3_TEXT`, I think the code flow will be like this.  
Firstly, it runs into the code block of `if (t == SQLITE3_TEXT)` (line 188) and set `pBlob = aBlob;` (line 192).  
Next, it runs into this block, and then return NULL immediately.  
Is that correct behavior?  
I think the correct source code should be like this.  
```
if (t == SQLITE_BLOB)   --> Line 184
...
else if (t == SQLITE3_TEXT)   --> Line 188
...
else   --> Line 199
{
   sqlite3_result_null(context);
   return;
}
```

mkgrgis replied:
The logic was simplifyed, @son-phamngoc 
1. SQLITE_BLOB => use original BLOB
2. SQLITE3_TEXT => use BLOB from normalized string, if no normal string `NULL`
3. Otherwise `NULL`
4.  If "no `NULL`" branch, use BLOB (`sqlite3UuidBlobToStr(pBlob, zs);`)

This function was not intuitive. Is it better now?

son-phamngoc replied:
Thanks for fixing. It looks better. I confirmed. Please help me to resolve this comment.

son-phamngoc commented on sqlite_fdw.c:
```diff
@@ -5738,6 +5739,8 @@ sqlite_affinity_eqv_to_pgtype(Oid type)
 	}
 }
 
+static const char *azType[] = { "?", "integer", "real", "text", "blob", "null" };
```
Const should be declared at the beginning of the file. Could you move this array to there?

mkgrgis replied:
Moved after declaration of functions. Please verify @son-phamngoc !

son-phamngoc replied:
Thanks for fixing. I confirmed. Please help me to resolve this comment.

son-phamngoc commented on deparse.c:
```diff
@@ -2366,8 +2403,27 @@ sqlite_deparse_direct_update_sql(StringInfo buf, PlannerInfo *root,
+			if (pg_attyp == UUIDOID && preferred_affinity == SQLITE3_TEXT)
+				appendStringInfo(buf, "sqlite_fdw_uuid_str(");
+			if (pg_attyp == TIMESTAMPOID && preferred_affinity == SQLITE_INTEGER)
+				appendStringInfo(buf, "strftime(");
```
I don't see any test case for this. Could you add it?

mkgrgis replied:
There is such TCs, @son-phamngoc 
https://github.com/mkgrgis/sqlite_fdw/blob/8cf4145ef00cb0d344a5a63b2ffa2528ea7d9120/sql/16.0/extra/uuid.sql#L128C1-L138C1

For `timestamp` this fixes untested error and I'll plan special PR about `timestamp` mixed affinity data type. So TC about `timestamp` to extend existed TCs will be in this PR.

son-phamngoc replied:
I saw your test cases about UUID, but I did not see test cases about timestamp, so I added this comment.  
Do you mean you are going to add those test cases in another PR?  
If so, I think it is better to remove the source code about timestamp from this PR.  
You can add it again in your PR about timestamp. It will be more consistent.  
In this PR, the source code about timestamp has no test to verify, so we cannot allow to merge it. Please consider.

mkgrgis replied:
@son-phamngoc , I have changed some lines around of `pg_attyp == TIMESTAMPOID && preferred_affinity == SQLITE_INTEGER` to `TODO` comment in https://github.com/pgspider/sqlite_fdw/pull/95/commits/bd75d245bdd74b649805cad193cf3375af828dfd , please verify. Also please notice UPDATE with `timestamp` with `column_type` was untested and not correct and still leave untested and not correct.

son-phamngoc replied:
Could you please remove those code completely from this PR?
The reasons:
- We would like to keep the code in master branch clean, with only official code (and comment) that has been tested and verified. No redundant comment is allowed to merge to it.
- The original purpose of this PR is for UUID only. You also have a plan to release a separated PR for `timestamp`. Therefore, it is better to keep only code (and also comment) about `UUID` only in this PR. You can move the code for `timestamp` to your new PR.

mkgrgis replied:
Thanks for clarification, @son-phamngoc ! Changed in https://github.com/pgspider/sqlite_fdw/pull/95/commits/fcb8b8aca94bd6628a524aa234cea7d7f5c05834.

son-phamngoc replied:
Thanks for fixing. I confirmed. Please help me to resolve this comment.

son-phamngoc commented on expected/16.0/extra/uuid.out:
```diff
@@ -362,153 +362,170 @@ SELECT * FROM "type_UUID+";
+--Testcase 104:
+ALTER FOREIGN TABLE "type_UUIDpk" ALTER COLUMN col OPTIONS (ADD column_type 'BLOB');
+--Testcase 105: NO ERR, but the same semantics!
+INSERT INTO "type_UUIDpk" VALUES ('{b0eebc99-9c0b4ef8-bb6d6bb9-bd380a12}');
```
What do you mean `the same semantics`?  
Could you add SELECT test cases after each INSERT test case? It will help developer to verify result after INSERT easily.

mkgrgis replied:
> What do you mean `the same semantics`?

We can have in SQLite some UUID value as value with `text` affinity and with `blob` affinity. This values will be transformed to equal internal PostgreSQL 16 bytes. This means this values have the same semantics for PostgreSQL, but not for SQLite. SQLite allow unique index with same UUID values with different affinities, but PostgreSQL disallow. This TC is about "SQLite allow unique index with same UUID values".

Additional TCs will be added in the 2nd commit of this review.

son-phamngoc replied:
Thanks for your explanation. I understood.

> Additional TCs will be added in the 2nd commit of this review.

Could you add these test cases?

mkgrgis replied:
Added in https://github.com/pgspider/sqlite_fdw/pull/95/commits/7cc6fed2240795e80bec225803959a7beb9d1d00 , @son-phamngoc . Please verify. Unfortunately the main change in `sqlite_fdw.c` also was only in private draft, but restored in https://github.com/pgspider/sqlite_fdw/pull/95/commits/7cc6fed2240795e80bec225803959a7beb9d1d00.

mkgrgis replied:
Note: I am not familiar with `git`, this branch was not synchronized with private draft during missing HEAD commit selecting. Now all is ok after restoring main contribution in https://github.com/pgspider/sqlite_fdw/commit/7cc6fed2240795e80bec225803959a7beb9d1d00 and https://github.com/pgspider/sqlite_fdw/commit/9c81b0fa941c1e56b4c6badcb5820c16c6e5581b

son-phamngoc replied:
Thank you. It's OK now. Please help me to resolve this comment.

son-phamngoc reviewed (commented):
@mkgrgis Thanks for your support.
I reviewed you PR. Could you check my comments?

mkgrgis commented on deparse.c:
```diff
@@ -164,6 +164,7 @@ static void sqlite_get_relation_column_alias_ids(Var *node, RelOptInfo *foreignr
 static char *sqlite_quote_identifier(const char *s, char q);
 static bool sqlite_contain_immutable_functions_walker(Node *node, void *context);
 static bool sqlite_is_valid_type(Oid type);
+int preferred_sqlite_affinity (Oid relid, int varattno);
```
Fixed. Really this function is static. Please verify, @son-phamngoc .

son-phamngoc replied:
Thanks for fixing. I confirmed. Please help me to resolve this comment.

mkgrgis commented on deparse.c:
```diff
@@ -2296,6 +2300,35 @@ sqlite_deparse_update(StringInfo buf, PlannerInfo *root,
 
+/* Preferred SQLite affinity from "column_type" foreign column option
+ * SQLITE_NULL if no value or no normal value
+ */
```
Changed.

son-phamngoc replied:
Thanks for fixing. I confirmed. Please help me to resolve this comment.

mkgrgis commented on deparse.c:
```diff
@@ -2296,6 +2300,35 @@ sqlite_deparse_update(StringInfo buf, PlannerInfo *root,
+			coltype = defGetString(def);
+			break;
+		}
+		elog(DEBUG4, "column type = %s", coltype);
```
Thanks, @son-phamngoc ! This is my mistake. Fixed.

son-phamngoc replied:
Thanks for fixing. I confirmed. Please help me to resolve this comment.

mkgrgis commented on sqlite_fdw.c:
```diff
@@ -5764,6 +5766,28 @@ sqlite_datatype(int t)
+    /* if (strcasecmp(t, azType[5]) == 0)
+     * error value
+     */
+    return SQLITE_NULL;
```
Fixed. This was code from some web page about SQLite.

son-phamngoc replied:
Thanks for fixing. I confirmed. Please help me to resolve this comment.

mkgrgis commented on sqlite_fdw.c:
```diff
@@ -5764,6 +5766,28 @@ sqlite_datatype(int t)
+        return SQLITE_TEXT;
+    if (strcasecmp(t, azType[4]) == 0)
+        return SQLITE_BLOB;
+    /* if (strcasecmp(t, azType[5]) == 0)
```
Yes, I think here will be no problems. Fixed.

son-phamngoc replied:
Thanks for fixing. I confirmed. Please help me to resolve this comment.

mkgrgis commented on sqlite_data_norm.c:
```diff
@@ -142,21 +143,68 @@ sqlite_fdw_uuid_blob (const unsigned char* s0, unsigned char* Blob)
+	static const char hex_dig[] = "0123456789abcdef";
+	int i, k;
+	unsigned char x;
+	k = 0;
```
This have rewritten to earlier `gcc` style. Thanks, @son-phamngoc ! We have discussed about aspect of variable initialization in earlier `gcc` versions with @t-kataym.

son-phamngoc replied:
Could you initialize value for `k` at line 153, same as `i`?
```C
int i = 0, k=0x550;
```

mkgrgis replied:
Changed in https://github.com/pgspider/sqlite_fdw/pull/95/commits/bd75d245bdd74b649805cad193cf3375af828dfd, really this will be more understandable. Please confirm, @son-phamngoc.

mkgrgis commented on sqlite_data_norm.c:
```diff
@@ -142,21 +143,68 @@ sqlite_fdw_uuid_blob (const unsigned char* s0, unsigned char* Blob)
+	{
+		pBlob = sqlite3_value_blob(arg);
+	}
+	if (t == SQLITE3_TEXT)
```
Yes, @son-phamngoc ! In this context `else if` is better.

son-phamngoc replied:
Thanks for fixing. I confirmed. Please help me to resolve this comment.

mkgrgis commented on sqlite_data_norm.c:
Fixed, @son-phamngoc , all comments updated, something added. Thanks for review of `sqlite_data_norm.c`!

son-phamngoc replied:
Thanks for fixing. I confirmed. Please help me to resolve this comment.

mkgrgis commented on deparse.c:
```diff
@@ -2366,8 +2403,27 @@ sqlite_deparse_direct_update_sql(StringInfo buf, PlannerInfo *root,
+			elog(DEBUG3, "sqlite_fdw : aff %d\n", preferred_affinity);
+			if (pg_attyp == UUIDOID && preferred_affinity == SQLITE3_TEXT)
+				appendStringInfo(buf, "sqlite_fdw_uuid_str(");
+			if (pg_attyp == TIMESTAMPOID && preferred_affinity == SQLITE_INTEGER)
```
Thanks, @son-phamngoc , your variant is better for case of addition of some new data types. (`float`, `macaddr`, `timestamp` are planned as mixed affinity)

son-phamngoc replied:
Thanks for fixing. I confirmed. Please help me to resolve this comment.

son-phamngoc commented on README.md:
@mkgrgis Could you help me to fix this comment?

mkgrgis replied:
@son-phamngoc , are you about all of cases _dataypes_ instead of _data types_ or something else?

son-phamngoc replied:
Yes, please search and fix all places:  
dataypes -> data types
becasuse -> because
sensetive -> sensitive

mkgrgis replied:
Fixed in https://github.com/pgspider/sqlite_fdw/pull/95/commits/fcb8b8aca94bd6628a524aa234cea7d7f5c05834, @son-phamngoc Please verify.

son-phamngoc replied:
Thanks for fixing. I confirmed. Please help me to resolve this comment.

mkgrgis commented:
> I reviewed you PR. Could you check my comments?

All your comments are checked, @son-phamngoc ! Let's request the second review?

son-phamngoc commented:
@mkgrgis Thanks for your support.  
I executed test for this PR in my environment and they all passed.
I see no problem in this PR for now.
I will leave the decision to @t-kataym.

t-kataym commented:
@son-phamngoc Thank you for your confirmation.
@mkgrgis Thank you for your contribution.
I will merge the PR.

mkgrgis commented:
Thanks, @t-kataym! My next PR from https://github.com/pgspider/sqlite_fdw/issues/66 series will be about ∞ in `float` and `numeric` data, the branch is almost ready.

mkgrgis commented:
Mixed affinity support PRs continued with https://github.com/pgspider/sqlite_fdw/pull/97.
